src/Servers/IIS/IIS/test/IIS.FunctionalTests/Http3Tests.cs

```diff
@@ -0,0 +1,194 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Quic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Server.IIS.FunctionalTests;
+using Microsoft.AspNetCore.Server.IntegrationTesting.Common;
+using Microsoft.AspNetCore.Server.IntegrationTesting.IIS;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Net.Http.Headers;
+using Microsoft.Win32;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.IIS.FunctionalTests
+{
+    [MsQuicSupported]
+    [Http3Supported]
+    [Collection(IISHttpsTestSiteCollection.Name)]
+    public class Http3Tests
+    {
+        public Http3Tests(IISTestSiteFixture fixture)
+        {
+            var port = TestPortHelper.GetNextSSLPort();
+            fixture.DeploymentParameters.ApplicationBaseUriHint = $"https://localhost:{port}/";
+            fixture.DeploymentParameters.AddHttpsToServerConfig();
+            fixture.DeploymentParameters.SetWindowsAuth(false);
+            Fixture = fixture;
+        }
+
+        public IISTestSiteFixture Fixture { get; }
+
+        [ConditionalFact]
+        public async Task Http3_Direct()
+        {
+            var handler = new HttpClientHandler();
+            handler.MaxResponseHeadersLength = 128;
+            handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
+            using var client = new HttpClient(handler);
+            client.DefaultRequestVersion = HttpVersion.Version30;
+            client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;
+            var response = await client.GetAsync(Fixture.Client.BaseAddress.ToString() + "Http3_Direct");
+
+            response.EnsureSuccessStatusCode();
+            Assert.Equal(HttpVersion.Version30, response.Version);
+            Assert.Equal("HTTP/3", await response.Content.ReadAsStringAsync());
+        }
+
+        [ConditionalFact]
+        public async Task Http3_AltSvcHeader_UpgradeFromHttp1()
+        {
+            var address = Fixture.Client.BaseAddress.ToString() + "Http3_AltSvcHeader_UpgradeFromHttp1";
+
+            var altsvc = $@"h3="":{new Uri(address).Port}""";
+            var handler = new HttpClientHandler();
+            // Needed on CI, the IIS Express cert we use isn't trusted there.
+            handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
```

Comment:
Should we only add this conditionally, i.e. check if this is running on ci?

Reply:
*shrug* this is common across many of our server tests.

Reply:
Should we move this into a helper method rather than have it in every test then?